lib/insights/api/common/rbac/access.rb

```diff
@@ -5,32 +5,32 @@ module RBAC
         class Access
           attr_reader :acl
           DEFAULT_LIMIT = 500
-          def initialize(resource, verb)
-            @resource = resource
-            @verb     = verb
-            @regexp   = Regexp.new(":(#{Regexp.escape(@resource)}|\\*):(#{Regexp.escape(@verb)}|\\*)")
-            @app_name = ENV["APP_NAME"]
+          def initialize(app_name = ENV["APP_NAME"])
+            @app_name = app_name
           end
 
           def process
             Service.call(RBACApiClient::AccessApi) do |api|
-              @acl ||= Service.paginate(api, :get_principal_access, {:limit => DEFAULT_LIMIT}, @app_name).select do |item|
-                @regexp.match?(item.permission)
-              end
+              @acl ||= Service.paginate(api, :get_principal_access, {:limit => DEFAULT_LIMIT}, @app_name).to_a
             end
             self
           end
 
-          def accessible?
-            @acl.any?
+          def accessible?(resource, verb, app_name = ENV['APP_NAME'])
```

Comment:
`app_name` is already given by `initialize`

Reply:
Right, maybe these should be `def accessible?(resource, verb, app_name = @app_name)`

Reply:
@gmcculloug When we fetch permissions for catalog and approval in a single RBAC call we might have to set @app_name = '*'  or @app_name = '\*' or @app_name = ""

But when you call this function you are trying to fetch it for a specific  app, so the @app_name is not a good default in this case. So the app_name = ENV['app_name'] is a better default and it sets the default to the current app that is using it.

Reply:
To clear the confusion I have changed the app_name in the constructor to app_name_filter since it can be '*'

Reply:
@mkanoor I am confused. When `@acl` is created, they are already limited by `@app_name_filter`. What do you get if here you `app_name = '*'`?

Or is it the opposite that you want to set `@app_name_filter = '*'` first and narrow it down to `app_name = 'catalog'` here? By using '*', do you really want to fetch permissions from ALL apps? The list will be very long. It might be better to call the method twice for two selected apps.

Reply:
@bzwei 
The RBAC service stores all the permissions for a user across applications. There are many applications that use RBAC and a single user can have permissions from many applications.

The get_principal_access call to the RBAC service gives you back these access/permissions. This API takes a filter called application_name listed here


https://github.com/RedHatInsights/insights-rbac-api-client-ruby/blob/master/docs/AccessApi.md#get_principal_access

The first argument to get principal access call is the name of the application to filter the access by so for a single application you can fetch your access by using the ENV['APP_NAME'] which is a pretty good default for services that only need to fetch access list for a single app e.g approval.

For apps that need access for 2 apps like catalog, we need access for approval and catalog we would have to make either 2 calls or make a single call with '*' or '' and then locally filter out the permissions we need.

Approach 1:

  x = get_principal_access('catalog',..)
  y = get_principal_access('approval',..)

Approach 2:
z = get_principal_access('*',...)

Or
 Approach 3:
a = get_principal_access('catalog, approval',...)


the get_principal_access call is expensive so its beneficial to make a single call.

Once you have the access available locally we cache that for the single API call since we might have to traverse the result multiple times.

So the admin_scope?, group_scope?, user_scope? calls are dealing with the locally available data so we don't have to go back to RBAC on a per call basis.

So when we initialize a instance of the Access class we are asking for the filter which can be

'approval'
'approval, catalog',
''
'*'

This is not the value that you would use when you call the admin_scope?, group_scope?, user_scope?, that has to be specific for the app that you are interested in.

As far as apps that deal with a single service the app_filter is the same as the ENV['APP_NAME'] and you don't have to add an extra argument

But for apps that have to deal with permissions for 2 different apps the app_name argument becomes important and it cannot default to the app_name_filter.

Reply:
My concern is to get all back through '*'. Since we can select multiple apps `a = get_principal_access('catalog, approval',...)`, then yes, this is good. Thanks for the clarification.

Reply:
My concern is if someone initializes the object with `app_name_filter` set to a single app_name that is not the same as `ENV['APP_NAME']` then they are required to pass `app_name` to every method call, which seems odd.

Does that seem like a reasonable use case to handle?

I think we could support this in the initialize method by allowing a user to pass the `app_name` as a second optional parameter, and if that is nil (the default) we set it to `app_name_filter` if it does not include a separator or wildcard character.  Otherwise, it defaults to `ENV['APP_NAME']` .

Something like this:
```ruby
def initialize(app_name_filter = ENV["APP_NAME"], app_name = nil)
  @app_name_filter = app_name_filter

  @app_name = if app_name.present?
                app_name
              elseif !app_name_filter.include?('*') and !app_name_filter.include?(',')
                app_name_filter
              else
                ENV["APP_NAME"]
              end
end
```

Probably a better way to do this, but trying to get the idea across.

Reply:
I don't have a real use-case for this at the moment, so willing to ignore this for now and address later if we find a need, at which time we'll have more information to drive the changes.

In most cases the app is going to be just checking it's own permissions so this makes sense as is.

Reply:
'*' is logically a superset of app_name.
```
x = Insights::API::Common::RBAC::Access.new('catalog')
x.accessible?('workflow','read','approval')
```
The above example will always return `false` even if the login user has true read permission to workflow. It is a coding error, but the bug cannot be easily revealed.